### ISSUE TITLE
CakeExecuteScript access denied errors on mono/mac

### ISSUE THREAD

**Issue by Redth** [cake-build/cake#1247]
### What You Are Seeing?
Getting an access denied error
```
Error: ApplicationName='/Users/redth/Desktop/breakcake/tools/Cake/Cake.exe', CommandLine='"/Users/redth/Desktop/breakcake/two.cake"', CurrentDirectory='/Users/redth/Desktop/breakcake', Native error= Access denied
```

### What is Expected?
No errors, secondary script should execute successfully

### What version of Cake are you using?
0.16

### Are you running on a 32 or 64 bit system?
x64

### What environment are you running on?  Windows? Linux? Mac?
Mac 10.11, Mac 10.12

### Are you running on a CI Server?  If so, which one?
N/A

### How Did You Get This To Happen? (Steps to Reproduce)

build.cake:
```
Information ("Running: {0}", "build.cake");

// Run the script from the subfolder
CakeExecuteScript ("two.cake");
```

two.cake:
```
Information ("Running: {0}", "two.cake");
```

### Output Log

~~~sh
Redth:breakcake redth$ cake --verbosity=diagnostic
Module directory does not exist.
Analyzing build script...
Analyzing /Users/redth/Desktop/breakcake/build.cake...
Processing build script...
Creating script session...
Adding reference to System.Data.dll...
Adding reference to System.Xml.dll...
Adding reference to System.Xml.Linq.dll...
Adding reference to Cake.Core.dll...
Adding reference to Cake.Common.dll...
Adding reference to Cake.exe...
Importing namespace Cake.Common...
Importing namespace Cake.Common.Build...
Importing namespace Cake.Common.Build.AppVeyor...
Importing namespace Cake.Common.Build.AppVeyor.Data...
Importing namespace Cake.Common.Build.BitbucketPipelines...
Importing namespace Cake.Common.Build.BitbucketPipelines.Data...
Importing namespace Cake.Common.Build.Bitrise...
Importing namespace Cake.Common.Build.Bitrise.Data...
Importing namespace Cake.Common.Build.ContinuaCI...
Importing namespace Cake.Common.Build.ContinuaCI.Data...
Importing namespace Cake.Common.Build.Jenkins...
Importing namespace Cake.Common.Build.Jenkins.Data...
Importing namespace Cake.Common.Build.TravisCI...
Importing namespace Cake.Common.Build.TravisCI.Data...
Importing namespace Cake.Common.Diagnostics...
Importing namespace Cake.Common.IO...
Importing namespace Cake.Common.IO.Paths...
Importing namespace Cake.Common.Net...
Importing namespace Cake.Common.Security...
Importing namespace Cake.Common.Solution...
Importing namespace Cake.Common.Solution.Project...
Importing namespace Cake.Common.Solution.Project.Properties...
Importing namespace Cake.Common.Solution.Project.XmlDoc...
Importing namespace Cake.Common.Text...
Importing namespace Cake.Common.Tools...
Importing namespace Cake.Common.Tools.Cake...
Importing namespace Cake.Common.Tools.Chocolatey...
Importing namespace Cake.Common.Tools.Chocolatey.ApiKey...
Importing namespace Cake.Common.Tools.Chocolatey.Config...
Importing namespace Cake.Common.Tools.Chocolatey.Features...
Importing namespace Cake.Common.Tools.Chocolatey.Install...
Importing namespace Cake.Common.Tools.Chocolatey.Pack...
Importing namespace Cake.Common.Tools.Chocolatey.Pin...
Importing namespace Cake.Common.Tools.Chocolatey.Push...
Importing namespace Cake.Common.Tools.Chocolatey.Sources...
Importing namespace Cake.Common.Tools.Chocolatey.Upgrade...
Importing namespace Cake.Common.Tools.DNU...
Importing namespace Cake.Common.Tools.DNU.Build...
Importing namespace Cake.Common.Tools.DNU.Pack...
Importing namespace Cake.Common.Tools.DNU.Restore...
Importing namespace Cake.Common.Tools.DotCover...
Importing namespace Cake.Common.Tools.DotCover.Analyse...
Importing namespace Cake.Common.Tools.DotCover.Cover...
Importing namespace Cake.Common.Tools.DotNetCore...
Importing namespace Cake.Common.Tools.DotNetCore.Build...
Importing namespace Cake.Common.Tools.DotNetCore.Execute...
Importing namespace Cake.Common.Tools.DotNetCore.Pack...
Importing namespace Cake.Common.Tools.DotNetCore.Publish...
Importing namespace Cake.Common.Tools.DotNetCore.Restore...
Importing namespace Cake.Common.Tools.DotNetCore.Run...
Importing namespace Cake.Common.Tools.DotNetCore.Test...
Importing namespace Cake.Common.Tools.DupFinder...
Importing namespace Cake.Common.Tools.Fixie...
Importing namespace Cake.Common.Tools.GitLink...
Importing namespace Cake.Common.Tools.GitReleaseManager...
Importing namespace Cake.Common.Tools.GitReleaseManager.AddAssets...
Importing namespace Cake.Common.Tools.GitReleaseManager.Close...
Importing namespace Cake.Common.Tools.GitReleaseManager.Create...
Importing namespace Cake.Common.Tools.GitReleaseManager.Export...
Importing namespace Cake.Common.Tools.GitReleaseManager.Publish...
Importing namespace Cake.Common.Tools.GitReleaseNotes...
Importing namespace Cake.Common.Tools.GitVersion...
Importing namespace Cake.Common.Tools.ILMerge...
Importing namespace Cake.Common.Tools.ILRepack...
Importing namespace Cake.Common.Tools.InspectCode...
Importing namespace Cake.Common.Tools.MSBuild...
Importing namespace Cake.Common.Tools.MSTest...
Importing namespace Cake.Common.Tools.NSIS...
Importing namespace Cake.Common.Tools.NuGet...
Importing namespace Cake.Common.Tools.NuGet.Install...
Importing namespace Cake.Common.Tools.NuGet.Pack...
Importing namespace Cake.Common.Tools.NuGet.Push...
Importing namespace Cake.Common.Tools.NuGet.Restore...
Importing namespace Cake.Common.Tools.NuGet.SetApiKey...
Importing namespace Cake.Common.Tools.NuGet.SetProxy...
Importing namespace Cake.Common.Tools.NuGet.Sources...
Importing namespace Cake.Common.Tools.NuGet.Update...
Importing namespace Cake.Common.Tools.NUnit...
Importing namespace Cake.Common.Tools.OctopusDeploy...
Importing namespace Cake.Common.Tools.OpenCover...
Importing namespace Cake.Common.Tools.ReportGenerator...
Importing namespace Cake.Common.Tools.ReportUnit...
Importing namespace Cake.Common.Tools.Roundhouse...
Importing namespace Cake.Common.Tools.SignTool...
Importing namespace Cake.Common.Tools.SpecFlow...
Importing namespace Cake.Common.Tools.SpecFlow.StepDefinitionReport...
Importing namespace Cake.Common.Tools.SpecFlow.TestExecutionReport...
Importing namespace Cake.Common.Tools.TextTransform...
Importing namespace Cake.Common.Tools.VSTest...
Importing namespace Cake.Common.Tools.WiX...
Importing namespace Cake.Common.Tools.WiX.Heat...
Importing namespace Cake.Common.Tools.XBuild...
Importing namespace Cake.Common.Tools.XUnit...
Importing namespace Cake.Common.Xml...
Importing namespace Cake.Core...
Importing namespace Cake.Core.Diagnostics...
Importing namespace Cake.Core.IO...
Importing namespace Cake.Core.Scripting...
Importing namespace System...
Importing namespace System.Collections.Generic...
Importing namespace System.IO...
Importing namespace System.Linq...
Importing namespace System.Text...
Importing namespace System.Threading.Tasks...
Compiling build script...
Running: build.cake
Executing: /Users/redth/Desktop/breakcake/tools/Cake/Cake.exe "/Users/redth/Desktop/breakcake/two.cake"
Error: System.ComponentModel.Win32Exception (0x80004005): ApplicationName='/Users/redth/Desktop/breakcake/tools/Cake/Cake.exe', CommandLine='"/Users/redth/Desktop/breakcake/two.cake"', CurrentDirectory='/Users/redth/Desktop/breakcake', Native error= Access denied
  at System.Diagnostics.Process.StartWithCreateProcess (System.Diagnostics.ProcessStartInfo startInfo) [0x001ee] in <affe4060066c42de8cdd6027cdb92b56>:0
~~~

**Comment by devlead** [comment]
Regression seems to be in the 4.5 Cake.xe assembly created by .NET Core and that Mono can't auto identify these as .NET/Mono assemblies. Mono has no issue executing said assembly when frame

A workaround is to specify the runtime to use for execution
```cake
public static CakeSettings GetCakeSettings(ICakeContext context, IDictionary<string, string> arguments = null)
{
    var settings = new CakeSettings { Arguments = arguments };
    if (context.Environment.Runtime.IsCoreClr)
    {
        settings.ToolPath = FindToolInPath(context, context.IsRunningOnUnix() ? "dotnet" : "dotnet.exe");
        settings.ArgumentCustomization = args => "./tools/Cake.CoreCLR/Cake.dll " + args.Render();

    }
    else if (context.IsRunningOnUnix())
    {
        settings.ToolPath = FindToolInPath(context, "mono");
        settings.ArgumentCustomization = args => "./tools/Cake/Cake.exe " + args.Render();
    }
    return settings;
}

public static FilePath FindToolInPath(ICakeContext context, string tool)
{
    var pathEnv = context.EnvironmentVariable("PATH");
    if (string.IsNullOrEmpty(pathEnv)||string.IsNullOrEmpty(tool))
    {
        return tool;
    }
    var paths = pathEnv.Split(new []{context.IsRunningOnUnix() ? ':' : ';'},  StringSplitOptions.RemoveEmptyEntries);
    return paths.Select(
            path=>new DirectoryPath(path).CombineWithFilePath(tool)
        ).FirstOrDefault(filePath=>System.IO.File.Exists(filePath.FullPath));
}
```

But this would only solve that tool, potentially as more .NET tools start to utilize .NET the more tools will start to fail on Mono.
.NET Core doesn't have auto magic but requires runtime to be specified.

We could introduce some magic in Tool / ToolSettings base, feels like a real potential rabbit hole though.

What we could do is somekind of runtime / framework option on the base settings class which perhaps could have Native as default and i.e. .NET Core / Mono / Java as alternatives Or just a FrameworkToolPath.

Perhaps we could add something to the tool registration to infer/let user specify runtime.

Tool aliases could potentially be aware of supported runtimes got it's specific tool an register that.

But this needs some thought, regardless ability to specify a tool runtime would be a good addition, if we make it "easy" for us it could be an "RuntimeToolPath" if not null used as process.

**Comment by Redth** [comment]
Here's a bit more information gathered...

There’s a bug i filed about this: https://bugzilla.xamarin.com/show_bug.cgi?id=44937

Essentially the trouble area is: https://github.com/mono/mono/blob/master/mono/io-layer/processes.c#L500

In that if statement, `first_word` and `second_word` are == 0 so it returns true and exits.

Now, the question: is mono’s interpretation of what’s managed or not incorrect, or is dotnetcore emitting an assembly with an incorrect CLR header?

**Comment by devlead** [comment]
@Redth found a workaround, will PR shortly.

**Comment by devlead** [comment]
Fixed by  #1266

**Comment by devlead** [comment]
@Redth v0.16.2 is now available on [nuget](https://www.nuget.org/packages/Cake/0.16.2) which addresses this issue. Big thanks to you for finding this and @akoeplinger who helped us sort this.
